wolk/service/registration_service/RegistrationService.cpp

```diff
@@ -140,23 +150,105 @@ std::unique_ptr<ErrorMessage> RegistrationService::removeDevices(const std::stri
     {
         const auto errorMessage = "Failed to generate the outgoing message.";
         LOG(ERROR) << errorPrefix << " -> " << errorMessage;
-        return std::unique_ptr<ErrorMessage>{
-          new ErrorMessage{deviceKey, errorMessage, std::chrono::system_clock::now()}};
+        return false;
     }
 
     // Send the message out
     if (!m_connectivityService.publish(message))
     {
         const auto errorMessage = "Failed to send the outgoing message.";
         LOG(ERROR) << errorPrefix << " -> " << errorMessage;
-        return std::unique_ptr<ErrorMessage>{
-          new ErrorMessage{deviceKey, errorMessage, std::chrono::system_clock::now()}};
+        return false;
+    }
+    return true;
+}
+
+std::shared_ptr<std::vector<std::string>> RegistrationService::obtainChildren(const std::string& deviceKey,
+                                                                              std::chrono::milliseconds timeout)
+{
+    LOG(TRACE) << METHOD_INFO;
+    const auto errorPrefix = "Failed to obtain children";
+
+    // Parse the message
+    auto message =
+      std::shared_ptr<Message>{m_protocol.makeOutboundMessage(deviceKey, ChildrenSynchronizationRequestMessage{})};
+    if (message == nullptr)
+    {
+        LOG(ERROR) << errorPrefix << " -> Failed to generate outgoing `ChildrenSynchronizationRequestMessage`.";
+        return nullptr;
+    }
+
+    // Lock the mutex for the map, just so the response arrival can't be faster than us putting the query data into the
+    // map.
+    std::atomic_bool called{false};
+    auto list = std::make_shared<std::vector<std::string>>();
+    {
+        std::lock_guard<std::mutex> lock{m_childrenSyncDevicesMutex};
+        // Send out the message
+        if (!m_connectivityService.publish(message))
+        {
+            LOG(ERROR) << errorPrefix << " -> Failed to send the outgoing `ChildrenSynchronizationRequestMessage`.";
+            return nullptr;
+        }
+        auto weakList = std::weak_ptr<std::vector<std::string>>{list};
+        m_childrenSynchronizationCallbacks[deviceKey].push(
+          [this, &called, weakList](const std::vector<std::string>& children) {
+              if (auto childrenList = weakList.lock())
+              {
+                  for (const auto& child : children)
+                      childrenList->emplace_back(child);
+                  called = true;
```

Comment:
You still have an issue of this callback being called after this method returns and thus modifying called variable once its out of scope

Reply:
You may pop the callback if not being called before returning